Allow verisure locks to be configured with a default code

ahayworth opened the pull request:
## Description:

This PR allows a user to configure a Verisure lock with a "default" code. This code will be used to lock or unlock the device when the user does not supply a code.

In #18791 @talzac reported that the Google Assistant could not unlock their Verisure locks, because the locks expect a user code. Google does not currently provide a way for us to prompt for a code, so the next best thing seems to be allowing a user to supply which code to use - hence the default.

Alternatives were considered, but not selected:
- Storing the code with Google in device customData: this felt superfluous and also not privacy-friendly.
- Allowing the user to configure a default code via entity configuration in the google_assistant/cloud components: This allows for maximum flexibility, but also feels a little bit like overkill - it might not be worth building unless users ask for it.

It should be noted that the Verisure lock component is currently the _only_ one that requires a code. The rest, while accepting one, do nothing with it at all.

There are some smokescreen tests added here, but that's because I don't have a Verisure system and I'm not sure how to test it otherwise. We'll ask @talzac to test, if they're willing.

**Related issue (if applicable):**
fixes #18791

**Pull request in [home-assistant.io](https://github.com/home-assistant/home-assistant.io) with documentation (if applicable):**
https://github.com/home-assistant/home-assistant.io/pull/7696

## Example entry for `configuration.yaml` (if applicable):
```yaml
verisure:
  ...
  locks: true
  default_lock_code: 1234
```

## Checklist:
  - [?] The code change is tested and works locally.
  - [x] Local tests pass with `tox`. **Your PR cannot be merged unless tests pass**
  - [x] There is no commented out code in this PR.

If user exposed functionality or configuration variables are added/changed:
  - [x] Documentation added/updated in [home-assistant.io](https://github.com/home-assistant/home-assistant.io)

ahayworth committed:
Allow verisure locks to be configured with a default code

ahayworth committed:
linting fix

balloob commented on homeassistant/components/verisure.py:
```diff
@@ -52,6 +53,8 @@
         vol.Optional(CONF_GIID): cv.string,
         vol.Optional(CONF_HYDROMETERS, default=True): cv.boolean,
         vol.Optional(CONF_LOCKS, default=True): cv.boolean,
+        vol.Optional(CONF_DEFAULT_LOCK_CODE, default=None):
```
We should not do a default and then match against `cv.string`. Since you use `.get` to get it from the dictionary, it will already return `None` instead of raise `KeyError` (which would happen when referring to a non existing key with `config['non-existing']`

ahayworth commented on homeassistant/components/lock/verisure.py:
```diff
@@ -96,13 +98,25 @@ def unlock(self, **kwargs):
+
+        code = kwargs.get(ATTR_CODE, self._default_lock_code)
+        if code is None:
+            _LOGGER.error("Code required but none provided.")
```
No problem. Do we have a formal style guide for that kind of stuff yet? If so, I will go and read it :)

MartinHjelmare replied:
https://developers.home-assistant.io/docs/en/development_guidelines.html#log-messages

ahayworth replied:
Oh, never mind, I found it.

ahayworth committed:
PR feedback

ahayworth commented on tests/components/lock/test_verisure.py:
```diff
@@ -0,0 +1,60 @@
+                                       return_value=None)
+
+            test['method']()
+            lock.set_lock_state.assert_not_called()
```
I've been trying to understand what you meant about preventing typos ... did you mean something like the following?
```diff
         for test in tests:
-            lock.set_lock_state = Mock(name='set_lock_state',
-                                       return_value=None)
+            mock = Mock(name='set_lock_state', return_value=None)
+            lock.set_lock_state = mock
 
             test['method']()
-            lock.set_lock_state.assert_not_called()
+            assert mock.call_count == 0
 
             test['method'](code='12345')
-            lock.set_lock_state.assert_called_with('12345', test['state'])
+            assert mock.call_args == call('12345', test['state'])
```

Otherwise I'm not sure I get the point - typos are typos, and it's certainly hard to prevent them on something as magical as python's mocking. I can and will make the change however, just want to make sure I understand the rationale.

MartinHjelmare replied:
Yes, exactly like that. Now if we would misspell `mock.call_count` to eg `mock.call_cuont` the test would fail.

ahayworth replied:
Makes sense, and https://github.com/home-assistant/home-assistant/pull/18873/commits/c6c927a0857c3e0676ba19a4003aa045f32dbb39 has those changes.

ahayworth committed:
PR feedback - try harder to prevent future typos

A python mock is a magical thing, and will respond to basicaly
any method you call on it. It's somewhat better to assert against
an explicit variable named 'mock', rather than to assert on the
method name you wanted to mock... could prevent a typo from messing up
tests.

ahayworth committed:
PR feedback: convert tests to integration-style tests

Set up a fake verisure hub, stub out a _lot_ of calls, then test
after platform discovery and service calls.

It should be noted that we're overriding the `update()` calls in
these tests. This was done to prevent even further mocking of
the verisure hub's responses.

Hopefully, this'll be a foundation for people to write more tests.

ahayworth commented on tests/components/lock/test_verisure.py:
```diff
@@ -0,0 +1,60 @@
+        hub.config = NO_DEFAULT_LOCK_CODE_CONFIG
+
+        from homeassistant.components.lock.verisure import VerisureDoorlock
+        lock = VerisureDoorlock("lock")
```
I suspected you'd ask me to do that, and commit https://github.com/home-assistant/home-assistant/pull/18873/commits/4a21d935985eb36b29795d28ef0f76e74fea5be0 has those changes.

It's a lot of setup and mocking to get the platform this way, especially for the two small unit tests that I wrote. I'm not sure that it's valuable in this particular instance to set up a whole platform since there are no tests for any of the other platform's behavior. Feels like wasted cpu cycles to me.

MartinHjelmare commented on tests/components/lock/test_verisure.py:
```diff
@@ -1,60 +1,112 @@
+    for test in LOCK_CODE_TESTS:
+        with mock_hub(NO_DEFAULT_LOCK_CODE_CONFIG, test['start']):
+            mock = Mock(name='set_lock_state', return_value=None)
+            lock.set_lock_state = mock
```
Mock the corresponding method on `hub.session` instead.

MartinHjelmare commented on tests/components/lock/test_verisure.py:
```diff
@@ -0,0 +1,112 @@
+async def setup_verisure_locks(hass, config):
+    """Set up mock verisure locks."""
+    with mock_hub(config):
+        discovery.load_platform(hass, LOCK_DOMAIN, VERISURE_DOMAIN, {},
```
Set up the verisure component instead and let it call discovery.

MartinHjelmare commented on tests/components/lock/test_verisure.py:
```diff
@@ -0,0 +1,112 @@
+    await setup_verisure_locks(hass, NO_DEFAULT_LOCK_CODE_CONFIG)
+    lock = hass.data[LOCK_DOMAIN].get_entity('lock.door_lock')
+    # Don't actually update state machine
+    lock.update = Mock('update', return_value=None)
```
Why mock update? We want the lock to update the state machine.

ahayworth replied:
I did this in https://github.com/home-assistant/home-assistant/commit/46ad64e2111bfd6c2131df2cff9d1b221a4df5d6.

MartinHjelmare reviewed (commented):
We should avoid mocking things directly on the entity.

ahayworth committed:
more pr feedback

MartinHjelmare reviewed (approved):
Looks great!
